src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -110,32 +112,32 @@ public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, ImmutableA
 
         // PROTOTYPE(NullableReferenceTypes): Check we are not using this method on type references in
         // member signatures visible outside the assembly. Consider overriding, implementing, NoPIA embedding, etc.
+        // PROTOTYPE(NullableReferenceTypes): [Obsolete("Use explicit NonNullTypes context")]
         public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, bool? isNullableIfReferenceType)
         {
-            return Create(typeSymbol, isNullableIfReferenceType, ImmutableArray<CustomModifier>.Empty);
+            return Create(typeSymbol, nonNullTypes: IsNullableToNonNullTypes(isNullableIfReferenceType), isAnnotated: IsNullableToIsAnnotated(isNullableIfReferenceType), ImmutableArray<CustomModifier>.Empty);
         }
 
         // PROTOTYPE(NullableReferenceTypes): Check we are not using this method on type references in
         // member signatures visible outside the assembly. Consider overriding, implementing, NoPIA embedding, etc.
-        public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, bool? isNullableIfReferenceType, ImmutableArray<CustomModifier> customModifiers)
+        public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, bool nonNullTypes, bool isAnnotated, ImmutableArray<CustomModifier> customModifiers)
         {
             if (typeSymbol is null)
             {
                 return null;
             }
 
-            if (isNullableIfReferenceType == null && typeSymbol.TypeKind == TypeKind.TypeParameter)
+            if ((isAnnotated || !(typeSymbol is TypeParameterSymbol)) &&
```

Comment:
Thanks. I'm still having some trouble with this code though. There may be an opportunity to simplify/clarify.

First, I think this will be easier to read if you invert this `if`. If I got my De Morgan's right:
```
if ((!isAnnotated && (typeSymbol is TypeParameterSymbol)) ||
    (isAnnotated && typeSymbol.IsReferenceType && !typeSymbol.IsNullableType()))
{
    // T (leave unannotated)
    // string? (leave annotated)
    Debug.Assert(!typeSymbol.IsNullableType());
}
else
{
    // string, int, T? where T:class (leave annotation)
    // int?, T? where T:struct (add annotation)
    isAnnotated = typeSymbol.IsNullableType();
}
```

Then, it feels like we can remove `&& !typeSymbol.IsNullableType()` from that check (since `Nullable<>` is a value type).
 
Finally, assuming I understood the cases right, there may be a further simplification: leaving `isAnnotated` unchanged in the string, int and T? cases. Maybe we just end up with `if (typeSymbol.IsNullableType()) { isAnnotated = true; }` ?

Reply:
Great, thanks. I've made the first two changes and added the comments. The third change lead to some failures with unconstrained type parameters, and I've left a comment to investigate further.

---
In reply to: [203933322](https://github.com/dotnet/roslyn/pull/28687#discussion_r203933322) [](ancestors = 203933322)